File: src/main/kotlin/me/icebear03/splendidenchants/enchant/data/Target.java
package me.icebear03.splendidenchants.enchant.data;

import org.bukkit.Material;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class Target {
    public static HashMap<String,Target> targets=new HashMap<>();
    static {
        //TODO INITIALIZE from targets.yml
    }
    private String id;
    private String name;
    private int capability; //该种物品最大的附魔容纳量
    private List<Material> types=new ArrayList<>();

    private Target(String id, String name, int capability, List<String> types) {
        this.id = id;
        this.name = name;
        this.capability = capability;
        for(String string:types) {
            this.types.add(Material.valueOf(string));
        }

        targets.put(id,this);
    }

    public String name() {
        return name;
    }

    public boolean isIn(Material type){
        return types.contains(type);
    }

    public static int maxCapability(Material type){
        int ans=99 ;/*TODO max in config.yml*/
        for(Target target:targets.values()){
            if(target.types.contains(type)){
                ans=Math.min(ans,target.capability);
            }
        }
        return ans;
    }
}

File: src/main/kotlin/me/icebear03/splendidenchants/enchant/data/Target.kt
package me.icebear03.splendidenchants.enchant.data

import me.icebear03.splendidenchants.util.loadAndUpdate
import org.bukkit.Material
import org.bukkit.inventory.EquipmentSlot
import taboolib.common.platform.function.info
import taboolib.module.configuration.Configuration
import java.util.concurrent.ConcurrentHashMap

data class Target(
    val id: String,
    val name: String,
    val capability: Int,
    val activeSlots: Set<EquipmentSlot>,
    val typeNames: List<String>
) {

    val types = typeNames.map { Material.valueOf(it) }

    fun maxCapability(type: Material): Int {
        var ans = 99 /*TODO max in config.yml*/
        for (target in targets.values) {
            if (target.types.contains(type)) {
                ans = ans.coerceAtMost(target.capability)
            }
        }
        return ans
    }

    companion object {

        val targets = ConcurrentHashMap<String, Target>()

        fun initialize() {
            val targetConfig = Configuration.loadAndUpdate("enchants/target.yml", listOf()) // TODO: 白名单列表
            targetConfig
            targetConfig.getKeys(false).forEach {
                targets[it] = Target(
                    it,
                    targetConfig.getString("$it.name")!!,
                    targetConfig.getInt("$it.max"),
                    targetConfig.getStringList("$it.active_slots").map { s -> EquipmentSlot.valueOf(s) }.toSet(),
                    targetConfig.getStringList("$it.types")
                )
            }
            targets["unknown"] = Target("unknown", "未定义", 16, hashSetOf(), arrayListOf())
            info("调试信息：加载附魔对象成功，共${targets.size}种对象！")
        }

        fun fromIdOrName(idOrName: String): Target {
            if (targets[idOrName] != null)
                return targets[idOrName]!!
            for (target in targets.values) {
                if (target.name == idOrName)
                    return target
            }
            return targets["unknown"]!!
        }
    }
}
